### ISSUE TITLE
ENB: Use pyzo syntax coloring

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#365]
An ongoing Engineering Notebook post. See also #361, a list of pyzo-related enhancements. 

**Background**

Pyzo's syntax coloring looks to be more than 10x faster than Leo's. Pyzo uses token-based regex "parsers", actually tokenizers, to guide syntax coloring. Eventually, Leo will have to generate parsers for each language in leo/modes.

**Prototyping**

`leo/external/pyzo` now contains minimally revised pyzo sources. Changes made:

- Revised imports
- Removed code from various `__init__.py` files. pyzo/__init__ started pyzo itself!
- See git log for full details.

Added traces to pyzo itself. This is easy because pyzo always opens a console window.

**Comment by edreamleo** [comment]
**Configuration**

The pyzo configuration is proving "challenging" to adapt to Leo's environment. The pyzo scheme seems too clever by half.

pyzo classes "register" configuration settings by defining `_styleElements` ivars, which are lists of tuples. Alas, the main syntax-coloring settings don't have `_styleElements` ivars.

Instead, `Manager.getStyleElementDescriptionsForAllParsers` gets syntax-coloring ivars from all *instantiated* parsers, that is, `cls._parserInstances.values()`. At present, Leo's pyzo code doesn't instantiate parsers as pyzo does.  The solution may simply be to call `Manager._collectParsers` "by hand".

Happily, it's easy to trace both the original pyzo code and Leo's version of the pyzo code. Comparing the two sets of traces shows what is happening, and what still needs to happen.

An initial hack didn't completely work.  I added a "force" keyword to `CodeEditorBase.setStyle` to force the creation of dict entries, regardless of whether the had a corresponding `_styleElements` entries.  But that wasn't enough. A way will soon be found...

**Comment by edreamleo** [comment]
**Tricks**

Here are a few hacks that may eventually go away. I have no idea what changes will eventually be made to Leo's core, but for now Leo's core should remain strictly unchanged.

1. Overrode Highlighter.rehighlight as follows:
```python
def rehighlight(self, p=None):
    '''Leo override, allowing the 'p' keyword arg.'''
    QtGui.QSyntaxHighlighter.rehighlight(self)
```

2. Added a `colorer` ivar to the `LeoCodeEditor` class, inited to `g.NullObject`. The main testing code is:

```python
    codeEditor = LeoCodeEditor()
    h = codeEditor.leo_highlighter
    h.colorer = g.NullObject()
        # Disable code in Leo's core.
    c.frame.body.colorizer.highlighter = h
```
So Leo's code can test `c.frame.body.colorizer.highlighter.colorer` without crashing.

**Comment by edreamleo** [comment]
**To do list**

**1. Support `@language`**

Once the python syntax coloring works, the parser/tokenizer must be enhanced to support Leo directives, especially the `@language` directive.  In fact, this directive will probably have to switch parsers. 

This must be done without damaging the speed of the all-important `Highlighter.highlightBlock` method.  The main loop of this method is the following:
```python
tokens = list(parser.parseLine(line, previousState))
for token in tokens :
    if isinstance(token, BlockState):
        self.setCurrentBlockState(token.state)
    else:
        try:
            styleFormat = nameToFormat(token.name)
            charFormat = styleFormat.textCharFormat
        except KeyError:
            continue
        self.setFormat(token.start,token.end-token.start,charFormat)
```
where `nameToFormat` is a synonym for `self._codeEditor.getStyleElementFormat`, which in turn is a dict lookup.  At present, `token.name` is a string, such as the following:
```
Syntax.openparen
Syntax.string
Syntax.nonidentifier
Syntax.identifier
Syntax.closeparen
Syntax.nonidentifier
Syntax.comment
```
token.name might also indicate language, allowing different settings for different languages, without affecting the speed of `Highlighter.highlightBlock` in any way.

But the real issue is that the line:
```python
tokens = list(parser.parseLine(line, previousState))
```
depends on the parser in effect on the line.  So Leo's colorizer may have to add more info to the user block data. 

**Bottom line**: the new code may require substantial modification, but the existing pyzo code will provide a guiding light.

**2. Write a script to create parsers from `leo/modes/*.py`**

Something like the `jEdit2p` script, but outputting regex-based parsers.

**Comment by edreamleo** [comment]
Rev bdb7b2fa90 completes this project.  Please reopen if problems are found.
